format/materialized-view-spec.md

```diff
@@ -0,0 +1,132 @@
+ - this work for additional information regarding copyright ownership.
+ - The ASF licenses this file to You under the Apache License, Version 2.0
+ - (the "License"); you may not use this file except in compliance with
+ - the License.  You may obtain a copy of the License at
+ -
+ -   http://www.apache.org/licenses/LICENSE-2.0
+ -
+ - Unless required by applicable law or agreed to in writing, software
+ - distributed under the License is distributed on an "AS IS" BASIS,
+ - WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ - See the License for the specific language governing permissions and
+ - limitations under the License.
+ -->
+
+# Iceberg Materialized View Spec
+
+## Background and Motivation
+Iceberg views are a powerful tool to abstract complex queries and share them among different engines.
+However, such views are not materialized by default, which means that they are re-computed every time they are queried.
+This can be inefficient for complex queries that are computed frequently.
+Iceberg Materialized views are a way to store the results of an Iceberg view to reuse the computation in subsequent queries.
+
+## Goals 
+The goal of this spec is to define the metadata associated with materialized views in Iceberg.
+Such metadata allows creating and querying Iceberg materialized views across different engines.
+
+## Specification
+A materialized view is an Iceberg view with a respective Iceberg table that stores the results of the view query.
+An Iceberg view is considered a materialized view if it has the `materialized.view` property set to `true`.
+A materialized view must also reference the storage table identifier in its `materialized.view.storage.table` property.
+
+The specification for the materialized view properties on the view is as follows:
+| Property name                              | Description                                                             |
+|--------------------------------------------|-------------------------------------------------------------------------|
+| `materialized.view`                | This property is used to mark whether a view is a materialized view. If set to `true`, the view is treated as a materialized view.|
+| `materialized.view.storage.table`  | This property specifies the identifier of the storage table associated with the materialized view.|
+
+In addition to the properties on the view, the storage table associated with the materialized view has the following properties:
+
+| Property name                        | Description                                                                   |
+|--------------------------------------|-------------------------------------------------------------------------------|
+| `base.snapshot.[UUID]`       | These properties store the snapshot IDs of the base tables at the time the materialized view's data was last updated. Each property is prefixed with `base.snapshot.` followed by the UUID of the base table.|
```

Comment:
Same comment, in https://docs.google.com/document/d/1UnhldHhe3Grz8JBngwXPA6ZZord1xMedY5ukEhZYF-A/edit#heading=h.r4ge4rk4hbp5 by @jankaul, it was a struct called 'lineage' that has these fields.  I think that was the proposal instead of ad-hoc table properties?

Reply:
For reference it was:
On Table Snapshot Metadata:
Struct 'lineage'
<b style="font-weight:normal;" id="docs-internal-guid-b3fee190-7fff-a4dc-6d07-94457ba7c2cd"><div dir="ltr" style="margin-left:0pt;" align="left">
v1 | Field Name | Description
-- | -- | --
required | refresh-version-id | Version id of the materialized view when the refresh operation was performed.
required | source-tables | A List of `source-table` records.

</div></b>


Struct 'source-table'
<b style="font-weight:normal;" id="docs-internal-guid-810541bf-7fff-c8b9-5933-a2e78987ca1d"><div dir="ltr" style="margin-left:0pt;" align="left">
v1 | Field Name | Description
-- | -- | --
required | uuid | Uuid of the source table
required | identifier | A full identifier record containing catalog, namespace and table-name fields.
required | snapshot-id | Snapshot id of the source table when the last refresh operation was performed.

</div></b>

Reply:
> You could have a MV whose SQL references the same table at different branches or timestamps.

Do we want to support that in materialized views yet? This will start requiring understanding the SQL logic and reasoning about aliases, etc, which is a high expectation from the engine code and requires some standardization too. Further, what SQL dialects support this today?

Reply:
Most engines out there support the time travel clause so its pretty ubiquitous now.   

We may not want to support the same table at different snapshots in the MV query tree right now.  But I think using a struct for each record allows us to add properties in the future to disambiguate the tables (maybe sequence number, parent view, path to root or query tree depth).  

I regularly see MVs on views that are 100 of non-distinct tables/views.  It happens a lot when users build semantic layers out of views.

Reply:
100 on non-disitinct tables/views is fine as long as non of them has time travel. The will all share the same snapshot ID and one record can represent them all. I think the only way to truly distinguish is the alias in the query. I do not see why adding query tree depth helps (i.e., it can still lead to collisions). Sequence number does not work either because it has to be recognized/shared by all engines.

Reply:
I'm not sure I understand your alias idea.  Can you elaborate with an example?  Thanks

Reply:
I actually think that putting info like this on snapshot is a lot easier for time-travel, as that is what the time travel mechanism is.  I understand having the right the lineage properties is not critical for the integrity of data but I feel we should to have a consistent view of the storage-table.

I think that would be hard to get in sync if we use table properties.  Look at the many problems we have just getting right schema to be read in some of these scenarios (rollback and time travel)   https://github.com/apache/iceberg/issues/5591 (unfixed), https://github.com/apache/iceberg/pull/9131, https://github.com/apache/iceberg/pull/1508, etc.

Reply:
Also as @jankaul pointed out, Trino already is modeling it as snapshot properties: https://github.com/trinodb/trino/blob/212455d3e1d393f58cbc395d2b9da47ed8f23dd8/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java#L2915